src/main/java/uk/nhs/tis/trainee/notifications/service/ProgrammeMembershipService.java

```diff
@@ -69,4 +100,233 @@ public boolean isExcluded(ProgrammeMembership programmeMembership) {
+   *
+   * @param programmeMembership The programme membership.
+   * @throws SchedulerException if any one of the notification jobs could not be removed.
+   */
+  public void deleteNotifications(ProgrammeMembership programmeMembership)
+      throws SchedulerException {
+
+    for (NotificationType milestone : NotificationType.getProgrammeUpdateNotificationTypes()) {
+
+      String jobId = milestone.toString() + "-" + programmeMembership.getTisId();
+      removeNotification(jobId); //remove existing notification if it exists
+    }
+  }
+
+  /**
+   * Schedule a programme membership notification.
+   *
+   * @param jobId      The job id. This must be unique for programme membership and notification
+   *                   milestone.
+   * @param jobDataMap The map of job data.
+   * @param when       The date to schedule the notification to be sent.
+   * @throws SchedulerException if the job could not be scheduled.
+   */
+  private void scheduleNotification(String jobId, JobDataMap jobDataMap, Date when)
+      throws SchedulerException {
+    JobDetail job = newJob(NotificationService.class)
+        .withIdentity(jobId)
+        .usingJobData(jobDataMap)
+        .storeDurably(false)
+        .build();
+
+    Trigger trigger = newTrigger()
+        .withIdentity(TRIGGER_ID_PREFIX + jobId)
+        .startAt(when)
+        .build();
+
+    Date scheduledDate = scheduler.scheduleJob(job, trigger);
+    log.info("Notification for {} scheduled for {}", jobId, scheduledDate);
+  }
+
+  /**
+   * Remove a scheduled notification if it exists.
+   *
+   * @param jobId The job id key to remove.
+   * @throws SchedulerException if the scheduler failed in its duties (non-existent jobs do not
+   *                            trigger this exception).
+   */
+  private void removeNotification(String jobId) throws SchedulerException {
+    JobKey jobKey = new JobKey(jobId);
+    // Delete the job and unschedule its triggers.
+    // We do not simply remove the trigger, since a replacement job may have different data
+    // (e.g. programme name).
+    scheduler.deleteJob(jobKey);
+    log.info("Removed any stale notification scheduled for {}", jobId);
+  }
+
+  /**
+   * Get a future schedule for a notification from the programme start date and day offset.
+   *
+   * @param programmeStartDate The programme starting date.
+   * @param daysBeforeStart    The number of days prior to the programme start date.
+   * @return The notification scheduling date and time.
+   */
+  private Date getScheduleDate(LocalDate programmeStartDate, int daysBeforeStart) {
+    Date milestone;
+    LocalDate milestoneDate = programmeStartDate.minusDays(daysBeforeStart);
+    if (!milestoneDate.isAfter(LocalDate.now())) {
+      // 'Missed' milestones: schedule to be sent soon, but not immediately
+      // in case of human editing 'jitter'.
+      milestone = Date.from(Instant.now()
+          .plus(PAST_MILESTONE_SCHEDULE_DELAY_HOURS, ChronoUnit.HOURS));
```

Comment:
The reason for this is to cover scenarios like the following: the LO creates (or edits) a PM and sets the start date to 6 weeks in the future by mistake. They then correct the date to 12 weeks (for example). If notifications were sent immediately, the trainee would receive a 8-week notification, and then no follow--up correction. By delaying the initial notification by an hour, it allows the LO to make the correction, which would then unschedule the incorrect immediate 8-week notification and set it (hopefully correctly) for 4 weeks in the future